File: solutions/sep-14/ChefAndLeftRight-CHEFLR/cheflr.py
def leftright(n):
    c = 'r'
    num = 1
    unmodded_num = 1
    modofunmodnum = 1
    level = 1
    MODNUM = 1000000007
    diff = 1
    for i in range(1,n+1):
        num *= 2
        unmodded_num *= 2
        if ((level % 2) != 0): 
            # (c == 'l') is just 2 * num
            if (c == 'r'):
                num += 2; 
                unmodded_num += 2; 
        else:
            # even level: L = 2 * num - 1; R = 2 * num + 1;
            if (c == 'l'):
                num -= 1
            else:
                num += 1

            if (c == 'l'):
                unmodded_num -= 1
            else:
                unmodded_num += 1
            
        num %= MODNUM;
        modofunmodnum = unmodded_num % MODNUM;
        #print "Debug ", "i=", i, "c=", c, "num=", num, "unmodded_num=", unmodded_num, "modofunmod=", modofunmodnum, "diff=", modofunmodnum - num
        #if (c == 'l'):
        #    c = 'r'
        level += 1
    return (num, unmodded_num, modofunmodnum)
            
print leftright(100)
print leftright(1000)
print leftright(10000)
print leftright(100000)

File: solutions/sep-14/ChefAndLeftRight-CHEFLR/cheflr.cpp
#include <stdio.h>
#include <string.h>

typedef unsigned long long int ULLI;
const ULLI MODNUM = 1000000007;
const int MAX_LEN = 100010;

int main()
{
	char input[MAX_LEN] = {'\0'};
	int nCases = 0, level = 0, retVal = 0;
	ULLI num = 1ULL; 
	char c = ' ';
	scanf("%d", &nCases);
	scanf("%c", &c); // read the '\n'
	
	while(nCases--)
	{
		memset(input, '\0', sizeof(char) * MAX_LEN);
		level = 1;
		num   = 1ULL;
		
		scanf("%s", input);
		//printf("Debug input %s=", input);

		for (int i = 0; input[i] != '\0'; ++i) {
			c = input[i];

			++level;
			
			num = ((num % MODNUM) * 2ULL) % MODNUM;
			
			// even level L = 2 * num; R = 2 * num + 2;
			if ((level & 1) == 0) { 
				// (c == 'l') is just 2 * num
				if (c == 'r') {
					num += 2ULL; 
				}
			}
			else {
				// odd level: L = 2 * num - 1; R = 2 * num + 1;
				num = (c == 'l') ? ((num - 1ULL)) : ((num + 1ULL));
			}
			
			num %= MODNUM;
		}
		printf("%llu\n", num);
	}
	
	return 0;
}
